File: KFLA.Contract/Contracts/Competency.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace KFLA.Contract
{
    public class Competency
    {
        public int ID { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public string ClusterID { get; set; }
        public string FactorID { get; set; }
        public List<string> LessSkilled { get; set; }
        public List<string> Skilled { get; set; }
        public List<string> Talented { get; set; }
        public List<string> OverusedSkill { get; set; }

        private Cluster _cluster;
        public Cluster Cluster
        {
            get { return _cluster; }
            set
            {
                _cluster = value;
                ClusterID = value?.ID;
            }
        }

        private Factor _factor;
        public Factor Factor
        {
            get { return _factor; }
            set
            {
                _factor = value;
                FactorID = value?.ID;
            }
        }

        public List<Question> Questions { get; set; } = new List<Question>();

        public string Context { get; set; }

        public List<string> Quotes { get; set; } = new List<string>();

        public string Positioning { get; set; }

        public List<string> Causes { get; set; } = new List<string>();

        public List<(string Type, string CaseStudy)> CaseStudies { get; set; } = new List<(string Type, string CaseStudy)>();

        public List<Tip> Tips { get; set; } = new List<Tip>();

        public List<string> JobAssignments { get; set; } = new List<string>();

        public List<(string Statement, string Suggestion)> TimeToReflect { get; set; } = new List<(string Statement, string Suggestion)>();

        public List<string> LearnMore { get; set; } = new List<string>();

        public List<string> DeepDiveResources { get; set; } = new List<string>();
    }
}

File: KFLA/ClientApp/src/models/Competency.ts
﻿import { Question } from "./Question";
import { Cluster, ClusterJSON } from "./Cluster";
import { Factor, FactorJSON } from "./Factor";
import { observable, computed, action } from 'mobx';
import { Evaluation } from "./Evaluation";
import { Tip } from "./Tip";


export interface CaseStudy {
    Type: string;
    Case: string;
}

export interface TimeToReflect {
    Statement: string;
    Suggestion: string;
}

export interface CompetencyJSON {
    ID: number;
    Name: string;
    Description: string;
    ClusterID: string;
    FactorID: string;
    Cluster: ClusterJSON;
    Factor: FactorJSON;
    Questions: string[];

    LessSkilled: string[];
    Skilled: string[];
    Talented: string[];
    OverusedSkill: string[];

    Context: string;
    Quotes: string[];
    Positioning: string;
    Causes: string[];
    CaseStudy: CaseStudy;
    JobAssignments: string[];
    TimeToReflect: TimeToReflect[];
    LearnMore: string[];
    DeepDiveResources: string[];
    Tips: Tip[];
}

export class Competency {
    ID: number;
    Name: string;
    Description: string;
    ClusterID: string;
    FactorID: string;
    LessSkilled: string[];
    Skilled: string[];
    Talented: string[];
    OverusedSkill: string[];
    Cluster: Cluster;
    Factor: Factor;

    Context: string;
    Quotes: string[];
    Positioning: string;
    Causes: string[];
    CaseStudy: CaseStudy;
    JobAssignments: string[];
    TimeToReflect: TimeToReflect[];
    LearnMore: string[];
    DeepDiveResources: string[];
    Tips: Tip[];

    @observable Questions: Question[];
    @observable Evaluation: Evaluation;
    @observable IsSelected: boolean;

    @action toggleSelection() {
        this.IsSelected = !this.IsSelected;
    }

    @action evaluateCompetency(evaluation?: Evaluation) {
        if (this.Evaluation) {
            const index = this.Evaluation.Competencies.indexOf(this, 0);
            if (index > -1) {
                this.Evaluation.Competencies.splice(index, 1);
            }
        }
        this.Evaluation = evaluation;
        if (this.Evaluation) {
            evaluation.Competencies.push(this);
        }
    }

    @computed get selectedQuestions(): Question[] {
        return this.Questions.filter(question => question.IsSelected);
    }

    @computed get IsEvaluated(): boolean {
        return this.Evaluation != null;
    }

    static fromJSON(json: CompetencyJSON | string): Competency {
        if (typeof json === 'string') {
            // if it's a string, parse it first
            return JSON.parse(json, Competency.reviver);
        } else {
            // create an instance of the class
            let user = Object.create(Competency.prototype);
            // copy all the fields from the json object
            return Object.assign(user, json, {
                // convert fields that need converting
                Cluster: Cluster.fromJSON(json.Cluster),
                Factor: Factor.fromJSON(json.Factor),
                Questions: json.Questions.map(questionJSON => new Question(questionJSON)),
            });
        }
    }

    static reviver(key: string, value: any): any {
        return key === "" ? Competency.fromJSON(value) : value;
    }
}